### ISSUE TITLE
Default values for frozen attrs classes

### ISSUE THREAD

**Issue by markusschmaus** [textX/textX#263]
Base type attribute are initialized as `None` with the comment that default values can be specified via object processors.

https://github.com/textX/textX/blob/c8a49a2fb286f5ad3f63b5a0a0fb33b8a312849e/textx/metamodel.py#L460-L464

However for frozen attrs classes, object processors are run too late to actually change the value of the attribute, so there is no obvious way to provide a default value for such classes.

One way to solve this would be to filter out any `None` values when constructing the dictionary passed to `__init__` and use the standard mechanisms for default values.

https://github.com/textX/textX/blob/c8a49a2fb286f5ad3f63b5a0a0fb33b8a312849e/textx/model.py#L836-L840

would become

```python
                attrs = {k: v for k, v in attrs.items()
                         if (k in obj.__class__._tx_attrs
                         or k == 'parent')
                         and v is not None}
```

**Comment by igordejanovic** [comment]
User have full control over user classes so the obvious and preferred way to provide defaults is in the class constructor. For dynamically created textX classes we don't have other options but to use object processors.

**Comment by markusschmaus** [comment]
Sounds reasonable. The issue is that currently the defaults provided by the class constructor are not used because TextX sets missing properties explicitly to `None`. Thats why I suggest filtering out the `None` values for user classes, though this would be a change to a public API.

**Comment by igordejanovic** [comment]
Yeah, it would be a change to the way it worked before. I guess that documenting the preferred way of providing the defaults might be a better approach. Something like:

```python
class MyUserClass:
    def __init__(self, first_param, ...):
        self.fist_param = <default value> if first_param is None else first_param
....
```
This is a good practice anyway as the usage of default values directly in the parameter list can lead to bugs if mutable values are used.

**Comment by markusschmaus** [comment]
I think best practice is to use a custom sentinel value as default rather than `None`, so that setting a value to `None` does not get confused with using the default value.

I discovered this on my own when I was bitten by some subtle bugs during a project at work, but I have since then discovered that attrs uses the same pattern with `attr.NOTHING` as sentinel value. Though I would always recommend to create your own sentinel value rather than reusing the sentinel value of another project. I just found [this blog post](https://treyhunner.com/2019/03/unique-and-sentinel-values-in-python) which also recommends this design pattern.

It would be great if there was a way to ask textx to not set missing values to `None`, in order to create compatibility with attrs and enable the user to use their custom sentinel value for defaults.

**Comment by igordejanovic** [comment]
Sounds like a good idea. I think it would be relatively straightforward to implement it as an optional parameter to `model_from*` where the default value may be `None` so it stays backward compatible.

**Comment by igordejanovic** [comment]
OTOH, textX will only [set value to `None` if it is missing in the input](http://textx.github.io/textX/stable/metamodel/#auto-initialization-of-the-attributes) so the only benefit of providing sentinel value would be if user-classes are also instantiated by the user code outside of textX. But, if you think it would be usable I think it would be ok for textX to provide it.
